Add the Cirrus40 keyboard

schuay opened the pull request:
A 40% keyboard with focus on usable encoders. The PCB uses external pullups for encoder pins, hence the custom initialization logic.

https://github.com/schuay/cirrus40

<!--- Provide a general summary of your changes in the title above. -->

<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

schuay commented:
@drashna What is the remaining process to land this PR? Do you have more comments?

filterpaper commented on keyboards/cirrus40/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,105 @@
+        'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R',
+        'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R',
+             '*', '*', '*', '*', '*', '*', '*', '*'
+    );
```
```suggestion
#if defined(CHORDAL_HOLD)
const char chordal_hold_layout[MATRIX_ROWS][MATRIX_COLS] PROGMEM =
    LAYOUT(
        'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R',
        'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R',
        'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R',
             '*', '*', '*', '*', '*', '*', '*', '*'
    );
#endif
```

zvecr replied:
It should just be removed, as it's exactly as what would be generated from the keyboard.json.

schuay committed:
Add the Cirrus40 keyboard

A 40% keyboard with focus on usable encoders. The PCB uses external
pullups for encoder pins, hence the custom initialization logic.

https://github.com/schuay/cirrus40

waffle87 commented on keyboards/cirrus40/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,97 @@
+        KC_F11,   KC_F3,    KC_F2,    KC_F1,    XXXXXXX,       XXXXXXX,  KC_LSFT,  KC_LCTL,  KC_LALT,  KC_LGUI,
+        KC_F12,   KC_F6,    KC_F5,    KC_F4,    XXXXXXX,       XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,  XXX,
+                  _______,  KC_ESC,   KC_SPC,   KC_TAB,        _______,  _______,  _______,  _______
+    ),
```
```suggestion
    [kNavigation] = LAYOUT(
        XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,       KC_PGUP,  _______,  _______,  _______,  XXXXXXX,
        KC_LGUI,  KC_LALT,  KC_LCTL,  KC_LSFT,  XXXXXXX,       KC_PGDN,  KC_LEFT,  KC_DOWN,  KC_UP,    KC_RGHT,
        _______,  _______,  _______,  _______,  XXXXXXX,       _______,  KC_HOME,  _______,  _______,  KC_END,
                  _______,  _______,  _______,  _______,       KC_ENT,   KC_BSPC,  KC_DEL,   _______
    ),

    [kNumbers] = LAYOUT(
        QK_BOOT,  XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,       XXXXXXX,  KC_7,     KC_8,     KC_9,     XXXXXXX,
        KC_LGUI,  KC_LALT,  KC_LCTL,  KC_LSFT,  XXXXXXX,       S(KC_G),  KC_1,     KC_2,     KC_3,     KC_0,
        QK_RBT,   XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,       XXXXXXX,  KC_4,     KC_5,     KC_6,     _______,
                  _______,  _______,  _______,  _______,       KC_ENT,   KC_BSPC,  KC_DEL,   _______
    ),

    [kFnKeys] = LAYOUT(
        KC_F10,   KC_F9,    KC_F8,    KC_F7,    XXXXXXX,       XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,
        KC_F11,   KC_F3,    KC_F2,    KC_F1,    XXXXXXX,       XXXXXXX,  KC_LSFT,  KC_LCTL,  KC_LALT,  KC_LGUI,
        KC_F12,   KC_F6,    KC_F5,    KC_F4,    XXXXXXX,       XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,
                  _______,  KC_ESC,   KC_SPC,   KC_TAB,        _______,  _______,  _______,  _______
    ),
```

schuay commented on keyboards/cirrus40/keyboard.json:
```diff
@@ -0,0 +1,100 @@
+    }
+  },
+  "layouts": {
+    "LAYOUT": {
```
Thanks, I've fixed the overlap.

schuay commented:
@waffle87 @filterpaper @zvecr ping

schuay committed:
core: add host_last_mouse_buttons() shadow helper

Track the most recent buttons mask passed to host_mouse_send() and
expose it via host_last_mouse_buttons(). Mirrors host_last_system_usage()
and host_last_consumer_usage() in shape and motivation.

Consumers that synthesize out-of-band mouse activity (Raw HID adapters,
pointing-device button handlers) need the current HID-level button
state to stamp onto reports they generate from non-button signals
(motion, wheel). Without this, a click-and-drag whose press and motion
arrive on separate code paths sees buttons=0 on the motion report and
the drag drops mid-stream.

schuay committed:
core: extract update/read_source_layers_cache_default; weak-wrap

Move the bodies of update_source_layers_cache and read_source_layers_cache
into update_source_layers_cache_default and read_source_layers_cache_default,
and make the original symbols weak wrappers that delegate to the
defaults.

This lets out-of-tree code (e.g. community modules adding new KEYLOC_*
sentinels for additional keypos types) strong-override the wrappers to
add their own cache buffers for those sentinels, then delegate the
existing matrix and encoder cases back to the *_default sibling.

Same shape as get_chordal_hold_default() in action_tapping.c. Pure
refactor: no behavior change for code that doesn't override the
wrappers.
